IMOD/manpages/extractpieces.man
.na
.nh
.TH extractpieces 1 4.6.34 IMOD
.SH NAME
extractpieces - Extract piece coordinates from header of image file
.SH SYNOPSIS
extractpieces  [options]  input_image_file  output_piece_list_file
.SH DESCRIPTION
EXTRACTPIECES will extract piece coordinates from the header of
an image file, if they are present, and produce a file with those
coordinates (a piece list file).  In addition, it can extract coordinates
from an image metadata file produced by SerialEM, and can do so
automatically if the image file itself does not have piece coordinates.
The program can also extract coordinates stored as attributes in an HDF
file, or from an ".idoc" file saved by SerialEM.
.SH OPTIONS
Extractpieces uses the PIP package for input (see the manual page for
pip(1)). Thus, the following options can be specified either as command line
arguments (with the -) or one per line in a command file (without the -):
.TP
.B -input \fROR\fP -InputFile 	 \fIFile name\fR
Image file with header information to extract.  If this option is not
entered, the first non-option argument will be taken as the input file.
An image file name is required unless the \fB-other\fR option is entered.
.TP
.B -output \fROR\fP -OutputFile 	 \fIFile name\fR
Output file for header values.  If this option is not entered, the second
non-option argument will be taken as the output file.
.TP
.B -mdoc \fROR\fP -MdocMetadataFile
Obtain values from a metadata file named with the name of the input file
plus '.mdoc', instead of from the extended header of the image file.  A
file with this name is generated by SerialEM when an option is selected to
save extra information in a metadata file, or when the coordinates are too
large to fit in the extended header.  You do not supply the name
of the .mdoc file when you use this option.  As of IMOD 4.8.7, the program
will look for such a file automatically if the requested values are not
present in the image file header.  Thus, this option is no longer necessary
except for forcing the program to ignore values in the header and
access the '.mdoc' file first.
.TP
.B -other \fROR\fP -OtherMetadataFile 	 \fIFile name\fR
Obtain values from the metadata file with the given name, instead of from
the extended header of the image file.  With this option, no input image
file needs to be supplied; but if you do omit the input file, you must
specify the output file with \fB-output\fR instead of as a non-option
argument.
.TP
.B -help \fROR\fP -usage
Print help output
.TP
.B -StandardInput
Read parameter entries from standard input.
.P
If the program is started with no command line arguments, it reverts to
interactive input with the following entries:
.P
  Image file with piece list information in header
  Output file for piece coordinates
.P
.SH HISTORY
.nf
  Written by David Mastronarde, October 1999
.fi
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/flib/image/extractpieces.f90
!*************EXTRACTPIECES**********************************************
!
! EXTRACTPIECES will extract piece coordinates from the header of
! an image file, if they are present, and produce a file with those
! coordinates (a piece list file) .
!
! David Mastronarde, 1/2/00
!
program extractpieces
  implicit none
  integer*4 nxyz(3), mxyz(3), nz, maxPiece, maxExtra, numExtraBytes, nbytes, iflags
  real*4, allocatable :: array(:)
  integer*4, allocatable :: ixPiece(:), iyPiece(:), izPiece(:)
  integer*4 mode, numPieces, i, ierr, ifAddMdoc, indAdoc, iTypeAdoc, numSect
  integer*4 montage
  real*4 dmin, dmax, dmean
  logical useMdoc, hdfFile
  integer*4 iiuFileType, iiuRetAdocIndex, AdocSetCurrent, AdocGetImageMetaInfo
  !
  character*320 inFile, outFile, metaFile
  character*10 metaOrHDF /'metadata'/
  !
  equivalence (nz, nxyz(3))

  integer*4 numOptArg, numNonOptArg
  integer*4 PipGetString, AdocOpenImageMetadata
  integer*4 PipGetInOutFile, PipParseInput, PipGetBoolean, PipGetLogical
  integer numOptions
  parameter (numOptions = 5)
  character*(120 * numOptions) options(1)
  options(1) = &
      'input:InputFile:FN:Name of input image file@'// &
      'output:OutputFile:FN:Name of output piece list file@'// &
      'mdoc:MdocMetadataFile:B:Take coordinates from metadata file named'// &
      ' inputfile.mdoc (omit name)@'// &
      'other:OtherMetadataFile:FN:Other metadata file to take '// &
      'coordinates from (follow with name)@'// &
      'help:usage:B:Print help output'

  metaFile = ' '
  inFile = ' '
  useMdoc = .false.
  hdfFile = .false.
  nz = 0
  maxPiece = 0
  !
  call PipExitOnError(0, 'ERROR: EXTRACTPIECES - ')
  call PipAllowCommaDefaults(1)
  ierr = PipParseInput(options, numOptions, '@', numOptArg, numNonOptArg)
  if (PipGetBoolean('help', ierr) == 0) then
    call PipPrintHelp('extractpieces', 0, 1, 1)
    call exit(0)
  endif
  ifAddMdoc = PipGetString('OtherMetadataFile', metaFile)
  ierr = PipGetLogical('MdocMetadataFile', useMdoc)
  if (useMdoc .and. ifAddMdoc == 0) call exitError( &
      'You cannot enter both -mdoc and -other')
  if (PipGetInOutFile('InputFile', 1, 'Name of input image file', inFile) &
      .ne. 0 .and. ifAddMdoc .ne. 0) call exitError( &
      'You must enter either an input image file or a metadata file with -other')
  if (PipGetInOutFile('OutputFile', 2, 'Name of output piece list file', &
      outFile) .ne. 0) call exitError('No output file specified')
  call PipDone()
  !
  if (inFile .ne. ' ') then
    call imopen(1, inFile, 'RO')
    call irdhdr(1, nxyz, mxyz, mode, dmin, dmax, dmean)
    hdfFile = iiuFileType(1) == 5 .or. iiuFileType(1) == 7
    if (hdfFile .and. ifAddMdoc .ne. 0) then
      indAdoc = iiuRetAdocIndex(1, 0, 0)
      if (indAdoc <= 0) call exitError('Getting autodoc index for HDF or IDOC file')
      if (AdocSetCurrent(indAdoc) .ne. 0) call exitError( &
          'Setting autodoc structure of HDF or IDOC file as current autodoc')
      if (AdocGetImageMetaInfo(montage, numSect, iTypeAdoc) < 0) then
        print *,'This file does not have metadata about each section'
        call exit(0)
      endif
      metaOrHDF = 'HDF'
    endif
  endif
  !
  ! try to open a metadata file regardless, unless got one from HDF file
  if (metaFile == ' ') metaFile = inFile
  if (.not.hdfFile .or. ifAddMdoc == 0) &
      indAdoc = AdocOpenImageMetadata(metaFile, ifAddMdoc, montage, numSect, iTypeAdoc)
  if (indAdoc > 0 .and. inFile == ' ') nz = numSect
  numPieces = 0

  if (.not.useMdoc .and. ifAddMdoc .ne. 0 .and. .not.hdfFile) then
    !
    ! Get data from the image header
    call iiuRetNumExtended(1, numExtraBytes)
    call iiuRetExtendedType(1, nbytes, iflags)
    maxExtra = numExtraBytes + 1024
    maxPiece = nz + 1024
    allocate(array(maxExtra / 4), ixPiece(maxPiece), iyPiece(maxPiece), &
        izPiece(maxPiece), stat = ierr)
    call memoryError(ierr, 'arrays for extra header or piece data')
    call iiuRetExtendedData(1, numExtraBytes, array)
    call get_extra_header_pieces(array, numExtraBytes, nbytes, iflags, nz, &
        ixPiece, iyPiece, izPiece, numPieces, maxPiece)
    if (numPieces == 0) then
      print *,'There are no piece coordinates in this image file'
      if (indAdoc .ne. -2)  &
          print *,'Looking for piece coordinates in the associated metadata file'
    endif
  endif
  
  if (numPieces == 0) then
    !
    ! Or get data from the autodoc file
    !
    ! Thanks to a bug in SerialEM, Montage flag may be missing.  So just plow
    ! ahead regardless
    if (indAdoc > 0 .and. numSect == nz) then
      if (maxPiece == 0) then
        maxPiece = nz + 1024
        allocate(ixPiece(maxPiece), iyPiece(maxPiece), izPiece(maxPiece), stat = ierr)
        call memoryError(ierr, 'arrays for piece data')
      endif
      call get_metadata_pieces(indAdoc, iTypeAdoc, nz, ixPiece, iyPiece, &
          izPiece, maxPiece, numPieces)
    endif
    !
    ! Give lots of different error messages
    if (indAdoc < 0 .or. numPieces < nz) then
      if (indAdoc == -2) then
        if (useMdoc .or. ifAddMdoc == 0) print *,'The metadata file does not exist'
      else if (indAdoc == -3) then
        print *,'The autodoc file is not a recognized type of image metadata file'
      else if (indAdoc == -1) then
        print *,'There was an error opening or reading the metadata file'
      else if (numPieces > 0 .or. numSect .ne. nz) &
          then
        write(*,'(a,a,a)')'The ', trim(metaOrHDF),' file does not have piece '// &
            'coordinates for every image in the file'
      else
        write(*,'(a,a,a)')'There are no piece coordinates in the ', trim(metaOrHDF), &
            ' file'
      endif
    endif
  endif
  !
  ! It used to output whatever is there, even if it is short, so restore that 6/24/16
  if (numPieces > 0) then
    call dopen(1, outFile, 'new', 'f')
    write(1, '(2i9,i7)') (ixPiece(i), iyPiece(i), izPiece(i), i = 1, numPieces)
    close(1)
    print *,numPieces, ' piece coordinates output to file'
  endif

  call iiuClose(1)
  !
  call exit(0)
end program extractpieces

